File: JAVA/ATM_Design/src/main/java/org/ceo/states/DispensingCashState.java
package org.ceo.states;

import org.ceo.enums.ATMState;
import org.ceo.factories.CardManagerFactory;
import org.ceo.factories.StateFactory;
import org.ceo.models.ATM;
import org.ceo.models.Card;
import org.ceo.services.CardManagerService;
import org.ceo.services.CashDispenserService;
import org.ceo.services.CashDispenserServiceImpl;

public class DispensingCashState extends ATMStateMachine {
    private final ATM atm;
    private final CashDispenserService cashDispenserService;

    public DispensingCashState(ATM atm) {
        this.atm = atm;
        this.cashDispenserService = new CashDispenserServiceImpl();
    }

    @Override
    public int dispenseCash(Card card, double amount, int transactionId) {
        CardManagerService cardManagerService = CardManagerFactory.getCardManager(card.getNetworkType(), card.getCardType());
        boolean isTxnSuccessful = cardManagerService.doTransaction(card, amount, transactionId);
        if(isTxnSuccessful) {
            this.cashDispenserService.dispenseCash(this.atm, transactionId, amount);
            System.out.println("Trx Successful");
        } else {
            System.out.println("Something went Wrong");
        }

        this.atm.changeState(StateFactory.getState(ATMState.EJECTING_CARD, this.atm));
        return 0;
    }

    @Override
    public ATMState getState() {
        return ATMState.DISPENSING_CASH;
    }
}

File: C++/ATM-System/src/states/DispensingCashState.cpp
#include "../states/DispensingCashState.h"
#include "../models/ATM.h"
#include "../states/EjectingCardState.h"
#include <iostream>
#include <stdexcept>

DispensingCashState::DispensingCashState(ATM* atm) : atm(atm) {}


int DispensingCashState::dispenseCash(Card card, double amount, int transactionId) {
    bool isTxnSuccessful = true;  // stubbed backend call

    if (isTxnSuccessful) {
        std::cout << "Transaction Successful" << std::endl;
    } else {
        std::cout << "Something went wrong" << std::endl;
    }

    atm->changeState(new EjectingCardState(atm));
    return 0;
}

ATMState DispensingCashState::getState() {
    return ATMState::DISPENSING_CASH;
}
